server/core/src/logical_locking.cpp

```diff
@@ -0,0 +1,305 @@
+                rst::get_property(_data_id, replica_number, "data_status"),
+                _replica_number));
+
+            if (_replica_number != replica_number) {
+                rst::update(_data_id, replica_number,
+                    json{{"data_is_dirty", std::to_string(_replica_status)}});
+            }
+        }
+    } // set_replica_statuses
+
+    auto lock_impl(
+        const std::uint64_t  _data_id,
+        const int            _replica_number,
+        const ill::lock_type _lock_type) -> int
+    {
+        try {
+            insert_data_status(_data_id);
+
+            const auto lock_status = lock_type_to_repl_status_map.at(_lock_type);
+
+            // The target replica should have its state set as well, but in a special manner
+            if (ill::lock_type::write == _lock_type) {
+                // A replica opened for write should be set to intermediate
+                rst::update(_data_id, _replica_number,
+                    json{{"data_is_dirty", std::to_string(INTERMEDIATE_REPLICA)}});
+            }
+            else {
+                // Even the target replica enters read lock state, so set that here
+                rst::update(_data_id, _replica_number,
+                    json{{"data_is_dirty", std::to_string(lock_status)}});
+            }
+
+            set_replica_statuses(_data_id, _replica_number, lock_status);
+
+            return 0;
+        }
+        catch (const irods::exception& e) {
+            irods::log(LOG_ERROR, fmt::format("[{}:{}] - [{}]", __FUNCTION__, __LINE__, e.what()));
+            return e.code();
+        }
+        catch (const std::exception& e) {
+            irods::log(LOG_ERROR, fmt::format("[{}:{}] - [{}]", __FUNCTION__, __LINE__, e.what()));
+            return SYS_INTERNAL_ERR;
+        }
+        catch (...) {
+            irods::log(LOG_ERROR, fmt::format("[{}:{}] - unknown error occurred", __FUNCTION__, __LINE__));
+            return SYS_UNKNOWN_ERROR;
+        }
+    } // lock_impl
+
+    auto unlock_impl(
+        const std::uint64_t _data_id,
+        const int           _replica_number,
+        const int           _replica_status,
+        const int           _other_replica_statuses) -> int
+    {
+        try {
+            auto replica_status = _replica_status;
+
+            if (ill::restore_status == _replica_status) {
+                if (const auto original_status = ill::get_original_replica_status(_data_id, _replica_number); -1 == original_status) {
+                    irods::log(LOG_ERROR, fmt::format(
+                        "[{}:{}] - failed to restore status for replica "
+                        "because no original status found in data_status column. "
+                        "Setting replica status to stale. "
+                        "[data_id=[{}], repl num=[{}]]",
+                        __FUNCTION__, __LINE__, _data_id, _replica_number));
+
+                    // TODO: Should this be stale or keep it locked for posterity?
```

Comment:
we've got data, we can't vouch for it... i think leave it stale.

Reply:
Okey dokey. Deleting comment.